File: macos_app/NagarikPatro/CalendarCore/Sources/CalendarCore/CalendarEventStore.swift
import Foundation

/// Loads and queries event and auspicious day data from JSON files.
public final class CalendarEventStore: Sendable {
    public static let shared = CalendarEventStore()

    /// Key: "YYYY-MM" (e.g., "2081-01"). Value: array of day event info.
    private let events: [String: [DayEventInfo]]
    /// Key: "YYYY-MM". Value: auspicious day info.
    private let auspicious: [String: AuspiciousMonth]

    private init() {
        self.events = Self.loadEvents()
        self.auspicious = Self.loadAuspicious()
    }

    // MARK: - Public API

    /// Get events for a specific day. Returns empty array if no data.
    public func events(year: Int, month: Int, day: Int) -> DayEventInfo? {
        let key = Self.monthKey(year: year, month: month)
        return events[key]?.first(where: { $0.day == day })
    }

    /// Get all events for a month.
    public func eventsForMonth(year: Int, month: Int) -> [DayEventInfo] {
        let key = Self.monthKey(year: year, month: month)
        return events[key] ?? []
    }

    /// Get auspicious data for a month.
    public func auspiciousForMonth(year: Int, month: Int) -> AuspiciousMonth? {
        let key = Self.monthKey(year: year, month: month)
        return auspicious[key]
    }

    /// Check if a specific day is auspicious for weddings.
    public func isAuspiciousWedding(year: Int, month: Int, day: Int) -> Bool {
        auspiciousForMonth(year: year, month: month)?.bibahaLagan.contains(day) ?? false
    }

    /// Check if a specific day is auspicious for bratabandha.
    public func isAuspiciousBratabandha(year: Int, month: Int, day: Int) -> Bool {
        auspiciousForMonth(year: year, month: month)?.bratabandha.contains(day) ?? false
    }

    /// Check if a specific day is auspicious for pasni.
    public func isAuspiciousPasni(year: Int, month: Int, day: Int) -> Bool {
        auspiciousForMonth(year: year, month: month)?.pasni.contains(day) ?? false
    }

    // MARK: - Models

    public struct DayEventInfo: Sendable {
        public let day: Int
        public let events: [String]
        public let eventsNp: [String]
        public let isHoliday: Bool
    }

    public struct AuspiciousMonth: Sendable {
        public let year: Int
        public let month: Int
        public let bibahaLagan: [Int]
        public let bratabandha: [Int]
        public let pasni: [Int]
    }

    // MARK: - Loading

    private static func monthKey(year: Int, month: Int) -> String {
        String(format: "%d-%02d", year, month)
    }

    private static func loadEvents() -> [String: [DayEventInfo]] {
        guard let url = Bundle.module.url(forResource: "nepali_calendar_events", withExtension: "json") else {
            print("WARNING: nepali_calendar_events.json not found")
            return [:]
        }
        do {
            let data = try Data(contentsOf: url)
            let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]
            var result: [String: [DayEventInfo]] = [:]
            result.reserveCapacity(json.count)

            for (key, value) in json {
                guard let monthData = value as? [String: Any],
                      let days = monthData["days"] as? [[String: Any]] else {
                    continue
                }
                result[key] = days.compactMap { dayJson -> DayEventInfo? in
                    guard let day = dayJson["day"] as? Int else { return nil }
                    return DayEventInfo(
                        day: day,
                        events: dayJson["events"] as? [String] ?? [],
                        eventsNp: dayJson["events_np"] as? [String] ?? [],
                        isHoliday: dayJson["is_holiday"] as? Bool ?? false
                    )
                }
            }
            return result
        } catch {
            print("ERROR: Failed to load events: \(error)")
            return [:]
        }
    }

    private static func loadAuspicious() -> [String: AuspiciousMonth] {
        guard let url = Bundle.module.url(forResource: "nepali_calendar_auspicious", withExtension: "json") else {
            print("WARNING: nepali_calendar_auspicious.json not found")
            return [:]
        }
        do {
            let data = try Data(contentsOf: url)
            let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]
            var result: [String: AuspiciousMonth] = [:]
            result.reserveCapacity(json.count)

            for (key, value) in json {
                guard let monthData = value as? [String: Any],
                      let year = monthData["year"] as? Int,
                      let month = monthData["month"] as? Int else {
                    continue
                }
                result[key] = AuspiciousMonth(
                    year: year,
                    month: month,
                    bibahaLagan: monthData["bibaha_lagan"] as? [Int] ?? [],
                    bratabandha: monthData["bratabandha"] as? [Int] ?? [],
                    pasni: monthData["pasni"] as? [Int] ?? []
                )
            }
            return result
        } catch {
            print("ERROR: Failed to load auspicious data: \(error)")
            return [:]
        }
    }
}

File: windows_app/NagarikPatro/Core/CalendarEventStore.cs
using System.Text.Json;

namespace NagarikPatro.Core;

/// <summary>
/// Loads and queries event and auspicious day data from JSON files.
/// </summary>
public sealed class CalendarEventStore
{
    private static readonly Lazy<CalendarEventStore> _instance = new(() => new CalendarEventStore());
    public static CalendarEventStore Instance => _instance.Value;

    private readonly Dictionary<string, List<DayEventInfo>> _events;
    private readonly Dictionary<string, AuspiciousMonth> _auspicious;

    private CalendarEventStore()
    {
        _events = LoadEvents();
        _auspicious = LoadAuspicious();
    }

    public IReadOnlyList<DayEventInfo> EventsForMonth(int year, int month)
    {
        var key = MonthKey(year, month);
        return _events.TryGetValue(key, out var events) ? events : [];
    }

    public DayEventInfo? EventsForDay(int year, int month, int day)
    {
        return EventsForMonth(year, month).FirstOrDefault(e => e.Day == day);
    }

    public AuspiciousMonth? AuspiciousForMonth(int year, int month)
    {
        var key = MonthKey(year, month);
        return _auspicious.TryGetValue(key, out var ausp) ? ausp : null;
    }

    public bool IsAuspiciousWedding(int year, int month, int day) =>
        AuspiciousForMonth(year, month)?.BibahaLagan.Contains(day) ?? false;

    public bool IsAuspiciousBratabandha(int year, int month, int day) =>
        AuspiciousForMonth(year, month)?.Bratabandha.Contains(day) ?? false;

    public bool IsAuspiciousPasni(int year, int month, int day) =>
        AuspiciousForMonth(year, month)?.Pasni.Contains(day) ?? false;

    // Models

    public sealed record DayEventInfo(
        int Day,
        IReadOnlyList<string> Events,
        IReadOnlyList<string> EventsNp,
        bool IsHoliday
    );

    public sealed record AuspiciousMonth(
        int Year,
        int Month,
        IReadOnlyList<int> BibahaLagan,
        IReadOnlyList<int> Bratabandha,
        IReadOnlyList<int> Pasni
    );

    // Loading

    private static string MonthKey(int year, int month) => $"{year}-{month:D2}";

    private static Dictionary<string, List<DayEventInfo>> LoadEvents()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "Data", "nepali_calendar_events.json");
        if (!File.Exists(path)) return [];

        try
        {
            var json = File.ReadAllText(path);
            using var doc = JsonDocument.Parse(json);
            var result = new Dictionary<string, List<DayEventInfo>>();

            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                if (!prop.Value.TryGetProperty("days", out var daysElement)) continue;

                var dayList = new List<DayEventInfo>();
                foreach (var dayEl in daysElement.EnumerateArray())
                {
                    if (!dayEl.TryGetProperty("day", out var dayProp)) continue;

                    dayList.Add(new DayEventInfo(
                        Day: dayProp.GetInt32(),
                        Events: ParseStringArray(dayEl, "events"),
                        EventsNp: ParseStringArray(dayEl, "events_np"),
                        IsHoliday: dayEl.TryGetProperty("is_holiday", out var h) && h.GetBoolean()
                    ));
                }
                result[prop.Name] = dayList;
            }
            return result;
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"Failed to load events: {ex.Message}");
            return [];
        }
    }

    private static Dictionary<string, AuspiciousMonth> LoadAuspicious()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "Data", "nepali_calendar_auspicious.json");
        if (!File.Exists(path)) return [];

        try
        {
            var json = File.ReadAllText(path);
            using var doc = JsonDocument.Parse(json);
            var result = new Dictionary<string, AuspiciousMonth>();

            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                var el = prop.Value;
                if (!el.TryGetProperty("year", out var yearProp) ||
                    !el.TryGetProperty("month", out var monthProp))
                    continue;

                result[prop.Name] = new AuspiciousMonth(
                    Year: yearProp.GetInt32(),
                    Month: monthProp.GetInt32(),
                    BibahaLagan: ParseIntArray(el, "bibaha_lagan"),
                    Bratabandha: ParseIntArray(el, "bratabandha"),
                    Pasni: ParseIntArray(el, "pasni")
                );
            }
            return result;
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"Failed to load auspicious: {ex.Message}");
            return [];
        }
    }

    private static List<string> ParseStringArray(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var arr)) return [];
        var list = new List<string>();
        foreach (var item in arr.EnumerateArray())
            list.Add(item.GetString() ?? "");
        return list;
    }

    private static List<int> ParseIntArray(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var arr)) return [];
        var list = new List<int>();
        foreach (var item in arr.EnumerateArray())
            list.Add(item.GetInt32());
        return list;
    }
}
